### PR TITLE
Adds options for deconstructing certain things without tools

### PR DESCRIPTION
**Pull request by DangerNoodle**
Recently while playing around, I noticed that placed anvils and other tools were trivial to construct, but required the full use of a hammer and screwdrivier to return to item form. After some work I went from a previous hacky method to adding a different style of deconstruction. This allows the option of a terrain/furniture feature yielding different items when deconstructed versus being dismantled sans tools. However, at present this capability is not made use of.

This applies to improvised shelters, sandbag barricades/walls, hay bales, the simple rope and pulley, straw beds, placed anvils, placed forges, placed stills, clay mounds, and sand mounds. For the last two, takedown recipes were added, allowing one to not require smashing them.

Similar ideas are implemented for More Survival Tools. Fire rings were changed to allow deconstruction. Blob terrain in Blazemod was also expanded in this way.

In general constructions requiring no tools to build were taken as the most obvious choice to implement this. Constructions requiring hammering but not involving nails (such as tarp tent, which could be taken down by hand) were added as well.

### CODE REVIEW

**Comment by DangerNoodle** [comment]
It would seem that @Blaze-whatdoyoumeanitsalreadytaken introduced formatting issues that is causing the error checker to cite an error 65. I am not sure how to resolve this.

**Comment by DangerNoodle** [comment]
I give up. I am undoing the changes to Vehicle Additions Pack in the hope that it will resolve the errors. This formatting is unusable and I am unable to fix it.

**Comment by DangerNoodle** [comment]
Not yet sure if the commit was formatted due to error-checker not handling it yet.

**Comment by Coolthulhu** [comment]
This is a hacky way of achieving it. Tool-less deconstruction should be controlled with a terrain flag, not by having a separate construction "recipe" per every type.

**Comment by DangerNoodle** [comment]
Not entirely certain how to handle that. The hack does allow showing the full list of what can be dismantled without tools, however.

What other methods would work?

**Comment by DangerNoodle** [comment]
I could possibly add a function similar to the takedown function for groundsheets and traps, but I do not know how to allow this to cite deconstruct to reset terrain and return multiple items.

**Comment by DangerNoodle** [comment]
Currently compile-testing a method of coding functions that allow checking for the presence of a flag, and allowing separate results for deconstructing versus dismantling. This allows adding only a single additional recipe to the construction menu instead of multiple additions, and allows a given terrain to yield different results depending on deconstruction versus dismantling.

Currently everything I am adding dismantling to will not have a deconstruct option as well, but th option will be there for future use.

**Comment by DangerNoodle** [comment]
There we go, apologies for the considerable delays this brought about.

**Comment by DangerNoodle** [comment]
Oh. Unmatched context, I would need the linter to consider "dismantle" valid.

**Comment by DangerNoodle** [comment]
This does present a problem in that I do not know if updating the linter in this pull request would be permitted, nor do I know whether it would cause the commit to succeed.

**Comment by mugling** [comment]
> This is something I would want to ask @mugling about if possible.

Lets walk through it then

File: src/construction.cpp [hunk 1]
```diff
@@ -771,6 +773,18 @@ bool construct::check_deconstruct( const tripoint &p )
...
+bool construct::check_dismantle( const tripoint &p )
+{
+    if( !g->m.has_flag( "DISMANTLE", p ) ) {
+        return false;
+    }
+    if( g->m.has_furn( p.x, p.y ) ) {
+        return g->m.furn( p.x, p.y ).obj().dismantle.can_do;
+    }
+    // terrain can only be taken down when there is no furniture in the way
+    return g->m.ter( p.x, p.y ).obj().dismantle.can_do;
```
Comment (Coolthulhu) [inline src/construction.cpp]: All the `p.x, p.y` here should be replaced with just `p`.

**Comment by DangerNoodle** [comment]
Ah, sorry. I would hope that my effort to update the permitted contexts were done correctly?

File: src/construction.cpp (hunk 1 above)
Comment (DangerNoodle) [inline src/construction.cpp]: That would be from my copying the original function, it seems.

File: src/mapdata.h [hunk 2]
```diff
@@ -196,6 +209,7 @@ struct map_data_common_t {
 
     map_bash_info        bash;
     map_deconstruct_info deconstruct;
+    map_dismantle_info   dismantle;
```
Comment (Coolthulhu) [inline src/mapdata.h]: Why have two separate blocks?
Why not just have a flag allow deconstructing without tools?

File: src/mapdata.h [hunk 3]
```diff
@@ -77,6 +77,18 @@ struct map_deconstruct_info {
     map_deconstruct_info() : can_do(false), deconstruct_above(false), ter_set(NULL_ID), furn_set(NULL_ID) {};
     bool load(JsonObject &jsobj, std::string member, bool is_furniture);
 };
+struct map_dismantle_info {
+    // Only if true, the terrain/furniture can be taken down
```
Comment (mugling) [inline src/mapdata.h]: Needs to be doxygen style comments eg. `/** summary here */`

File: src/mapdata.h (hunk 2 above)
Comment (DangerNoodle) [inline src/mapdata.h]: That would still have the problem of the required tools being on the JSON side of things, and making required tools optional in certain contexts would be even more of a headache that this was to implement, and be harder to convey to the player.

File: src/mapdata.h [hunk 4]
```diff
@@ -77,6 +77,18 @@ struct map_deconstruct_info {
...
+    bool can_do;
+    // In the unlikely event walls can be takn down without tools.
+    bool dismantle_above;
+    // items you get when dismantling.
+    std::string drop_group;
+    ter_str_id ter_set;    // terrain to set (REQUIRED for terrain))
+    furn_str_id furn_set;    // furniture to set (only used by furniture, not terrain)
+    map_dismantle_info() : can_do(false), dismantle_above(false), ter_set(NULL_ID), furn_set(NULL_ID) {};
```
Comment (mugling) [inline src/mapdata.h]: Please use C++11 default initialization. For example `bool dismantle_above = false`. You only need to retain the constructor for `NULL_ID`

File: src/mapdata.cpp [hunk 5]
```diff
@@ -970,10 +989,30 @@ void check_decon_items(const map_deconstruct_info &mbi, const std::string &id, b
...
+void check_takedown_items(const map_dismantle_info &mbi, const std::string &id, bool is_terrain)
+{
+    if (!mbi.can_do) {
+        return;
+    }
+    if( !item_group::group_is_defined( mbi.drop_group ) ) {
+        debugmsg( "%s: dismantling result item group %s does not exist", id.c_str(), mbi.drop_group.c_str() );
+    }
+    if (is_terrain && mbi.ter_set.is_empty()) { // Some tiles specify t_null explicitly
+        debugmsg("dismantling result terrain of %s is undefined/empty", id.c_str());
```
Comment (mugling) [inline src/mapdata.cpp]: Style:
```
if( is_terrain && mbi.ter_set.is_empty() ) {
    // Some tiles specify t_null explicitly
    debugmsg( "dismantling result terrain of %s is undefined/empty", id.c_str() );
}
```

Need to fix all other occurences

File: data/json/furniture.json [hunk 6]
```diff
@@ -96,8 +102,8 @@
         "bgcolor": "brown",
         "move_cost_mod": 3,
         "required_str": -1,
-        "flags": ["TRANSPARENT", "MOUNTABLE", "BLOCKSDOOR", "SHORT"],
-        "deconstruct": {
+        "flags": ["TRANSPARENT", "MOUNTABLE", "BLOCKSDOOR", "SHORT", "DISMANTLE"],
```
Comment (mugling) [inline data/json/furniture.json]: Needing a flag and a member object is a code smell

File: data/json/terrain.json [hunk 7]
```diff
@@ -4018,7 +4016,6 @@
         "color": "ltblue",
         "move_cost": 5,
         "flags": ["TRANSPARENT", "LIQUID", "SWIMMABLE", "SALT_WATER", "FISHABLE"],
-        "connects_to": "WATER",
```
Comment (mugling) [inline data/json/terrain.json]: Why this change?

File: tools/format/format.conf [hunk 8]
```diff
@@ -45,6 +45,9 @@
 ^(terrain|furniture):deconstruct:ter_set
 ^(terrain|furniture):deconstruct:furn_set
 ^(terrain|furniture):deconstruct:items
+^(terrain|furniture):dismantle:ter_set
```
Comment (mugling) [inline tools/format/format.conf]: This is fine

File: tools/format/format.conf [hunk 9]
```diff
@@ -358,19 +363,19 @@ mission_definition:destination
...
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:(item|group|distribution|collection)
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:prob
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:container-item
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:(damage|charges|count)=ARRAY,NOWRAP
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:(damage|charges|count)-min
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:(damage|charges|count)-max
+(^item_group|bash|deconstruct|dismantle|death_drops):.*(items|groups|entries|distribution|collection)
+(^item_group|bash|deconstruct|dismantle|death_drops):.*(items|groups|entries|distribution|collection):@=NOWRAP
+(^item_group|bash|deconstruct|dismantle|death_drops):.*(items|groups|entries|distribution|collection):@:(item|group|distribution|collection)
+(^item_group|bash|deconstruct|dismantle|dismantle|death_drops):.*(items|groups|entries|distribution|collection):@:prob
```
Comment (mugling) [inline tools/format/format.conf]: Duplicate entry

File: tools/format/format.conf [hunk 10]
```diff
@@ -358,19 +363,19 @@ mission_definition:destination
...
 # Item groups can appear in many contexts and with arbitrarily deep nesting
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection)
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@=NOWRAP
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:(item|group|distribution|collection)
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:prob
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:container-item
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:(damage|charges|count)=ARRAY,NOWRAP
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:(damage|charges|count)-min
-(^item_group|bash|deconstruct|death_drops):.*(items|groups|entries|distribution|collection):@:(damage|charges|count)-max
+(^item_group|bash|deconstruct|dismantle|death_drops):.*(items|groups|entries|distribution|collection)
```
Comment (mugling) [inline tools/format/format.conf]: May be overly permissive

**Comment by DangerNoodle** [comment]
At the moment I am afraid this is perhaps beyond my ability. My original method, while hacky, at least worked barring linting failures.

Apologies for this, but I do not think I can salvage this at present. Especially not while additionally being tripped up by problems in a second pull request, attempting to balance both it likely making it worse.

My failure to get used to the linting requirements is frustrating enough. I feel like the only one at the moment having such difficulty with correct formatting.

**Comment by mugling** [comment]
You need to start editing locally (not via github) and running the game binary after each edit. Your recent commits have all contained syntax errors which are unrelated to linting. You are guaranteed a `debugmsg` for invalid JSON.

**Comment by DangerNoodle** [comment]
I had done a compile and load test when I first started testing everything added to https://github.com/CleverRaven/Cataclysm-DDA/pull/18346/commits/4008929d3915b89001231e5791164778abbe0796, and it seemed to work as intended after a few early corrections (both in the source, and in JSON). It wasn't until after the commit failed that I started to rush through things in an attempt to pinpoint why it was not working in spite of a successful test.

**Comment by Coolthulhu** [comment]
I mean just a simple flag.

You could allow it like that:
* Things without flag need full deconstruction set
* Things with flag can be deconstructed OR taken down (both are fine), producing the same result

You don't need a separate deconstruction set just for things that can be taken down without tools.

**Comment by DangerNoodle** [comment]
I am unsure how to implement that, however.
